### DISCUSSION TITLE
RLS: Update with where clause checks select policy

### DISCUSSION THREAD

**Discussion by psteinroe** [supabase/supabase#9536] [Questions]
Hey,

I just encountered a weird behaviour or RLS policies and can't find a way around it. It seems like Postgres checks the `select` policy during an update if a `where` clause is used.

I need to achieve the following. A user can access a conversation if `conversation.inbox_id` points to an inbox the user can access. A user should now be able to update the inbox_id, hence moving it into another inbox. The user does not have access to the same conversation if it is in the new inbox, hence looses access to the very same conversation after updating the inbox_id.
In short:
- user can select the table entry
- user can update the table entry, and with that update the user cannot select the table entry anymore

Here is a simple test case that shows the issue:

```sql
begin;

select plan(2);

create table test (
  id integer primary key,
  val text
);

alter table test enable row level security;

create policy upd on test for update to authenticated 
using (true);

create policy sel on test for select to authenticated 
using (false);

insert into test (id, val) values (1, '1'), (2, '2');

set role authenticated;

prepare update_test as update test set val = 'Test' where id = 1;

select lives_ok('update_test', 'should be able to update'); -- passes

select is((select val from test where id = 1), 'Test', 'should update'); -- fails

select * from finish();

rollback;
```

If the select policy is changed to true, it works. I don't understand why this is happening, as I am only doing an update.

```sql
begin;

select plan(2);

create table test (
  id integer primary key,
  val text
);

alter table test enable row level security;

create policy upd on test for update to authenticated 
using (true);

create policy sel on test for select to authenticated 
using (true); -- set to true

insert into test (id, val) values (1, '1'), (2, '2');

set role authenticated;

prepare update_test as update test set val = 'Test' where id = 1;

select lives_ok('update_test', 'should be able to update'); -- passes

select is((select val from test where id = 1), 'Test', 'should update'); -- passes

select * from finish();

rollback;
```

Is this expected behaviour? If yes, what is the purpose of defining a `using` clause for the update?

**Comment by psteinroe** [comment]
@GaryAustin1 maybe also interesting for the rls research

**Comment by GaryAustin1** [comment]
Yep from the Postgres [docs](https://www.postgresql.org/docs/current/sql-createpolicy.html):
`Typically an UPDATE command also needs to read data from columns in the relation being updated (e.g., in a WHERE clause or a RETURNING clause, or in an expression on the right hand side of the SET clause). In this case, SELECT rights are also required on the relation being updated, and the appropriate SELECT or ALL policies will be applied in addition to the UPDATE policies. Thus the user must have access to the row(s) being updated through a SELECT or ALL policy in addition to being granted permission to update the row(s) via an UPDATE or ALL policy.`

So the same issues will exist as in just a select policy....

**Reply by psteinroe** [reply]
Thanks for the reference! I understand `Thus the user must have access to the row(s) being updated` as the user having access to the row **before** it is being updated, not after. I will post another test where access is granted to the row before the update, but not after.

**Reply by GaryAustin1** [reply]
Hey, I did not fully read your description just the select question.  There is a case with update that just can't be done if the RLS is referencing a change from the update.  I'll see if I can find it and if it matches what you are doing.

**Comment by psteinroe** [comment]
@GaryAustin1 here is an updated test that showcases the issue better

```sql
begin;

select plan(2);

create table test (
  id integer primary key,
  val text
);

alter table test enable row level security;

create policy upd on test for update to authenticated 
using (true); -- always allow update

create policy sel on test for select to authenticated 
using (val = '1');  -- allow select before update

insert into test (id, val) values (1, '1');

set role authenticated;

prepare update_test as update test set val = 'Test' where id = 1;

select lives_ok('update_test', 'should be able to update'); -- fails

select is((select val from test where id = 1), 'Test', 'should update'); -- fails

select * from finish();

rollback;
```

**Reply by psteinroe** [reply]
I assumed that as long as `returning` is not used, the select policy is irrelevant.

**Reply by GaryAustin1** [reply]
I've not been able to find the thread that I'm thinking of, but pretty sure your case is it.  It had something to do with marking thread hidden or deleted with the update not working....
This seems to be your case, but no answer...
https://stackoverflow.com/questions/69144319/unable-to-update-row-of-table-with-row-level-security

**Reply by psteinroe** [reply]
That doesn't make sense to me tbh. Why define an update using policy if the select is used anyways? I will ask in some Postgres forums. 👍🏼

**Reply by psteinroe** [reply]
Maybe @michelp can enlighten us yet again? 💚

**Reply by michelp** [reply]
The first issue is that you are getting permission denied for relation, which is always a GRANT related error, authenticated needs grants to access the table object itself.

The second issue is that the update implies a select, but you don't provide a WITH CHECK on the select policy for those cases, here's a working snippet:

```


create role authenticated;
    
drop table if exists test;

create table test (
  id integer primary key,
  val text
);

grant select, update on test to authenticated;

alter table test enable row level security;

create policy upd on test for update to authenticated
using (true);

create policy sel on test for select to authenticated
using (val = '1')
with check (true);  -- allow select before update

insert into test (id, val) values (1, '1');

set role authenticated;

prepare update_test as update test set val = 'Test' where id = 1;

execute update_test;
```

**Reply by GaryAustin1** [reply]
I've never seen a select policy with with check and it does not work when I run it...
![image](https://user-images.githubusercontent.com/54564956/195712298-f55e0816-2209-4abd-8408-b664d66df822.png)

If you meant with check(true) on the update policy, that is optional as it uses the using policy otherwise, and I checked that earlier just to make sure.

**Reply by michelp** [reply]
Ah my bad, I didn't catch the error, I've added a log function to show how the select policy is failing on the update because it is checked first, but the updated tuple has the new value.  This seems counter intuitive to me but not necessarily wrong, it seems like this would be the case where you with have a different WITH CHECK policy vs just a bare USING:

```
CREATE OR REPLACE function log(text) RETURNS boolean AS
$$
DECLARE
    r boolean;
BEGIN
  RAISE NOTICE '%', $1;
  RETURN true;
END;
$$ language plpgsql;

create role authenticated;
    
drop table if exists test;

create table test (
  id integer primary key,
  val text
);

grant select, update on test to authenticated;

alter table test enable row level security;

create policy upd on test for update to authenticated
using (true);

create policy sel on test for select to authenticated
using (log(val) and val = '1');

insert into test (id, val) values (1, '1');

set role authenticated;

prepare update_test as update test set val = 'Test' where id = 1;

execute update_test;
```

Returns:

```
CREATE FUNCTION
CREATE ROLE
NOTICE:  table "test" does not exist, skipping
DROP TABLE
CREATE TABLE
GRANT
ALTER TABLE
CREATE POLICY
CREATE POLICY
INSERT 0 1
SET
PREPARE
NOTICE:  1
NOTICE:  Test
ERROR:  new row violates row-level security policy for table "test"
```

**Reply by GaryAustin1** [reply]
It looks like you have to meet the select policy even on the updated row.  (I did add with check true and it makes no difference).

Reading [here](https://www.postgresql.org/docs/current/sql-createpolicy.html) at Application of Multiple Policies and looking at the chart above it, seems to be the case.  The new row has to be select-able to succeed...

**Reply by psteinroe** [reply]
As far as I can see, the case mentioned above is not possible then. From the docs it seems like pg doesn't make a difference between a where clause and a returning statement, despite the where being applied to the current row and the returning to the updated one only. 

_If read access is required to the existing or new row (for example, a WHERE or RETURNING clause that refers to columns from the relation)._ - footnote of table 286

I hope I miss something though.
